crates/ruff_db/src/vfs.rs

```diff
@@ -0,0 +1,318 @@
+use std::sync::Arc;
+
+use countme::Count;
+use dashmap::mapref::entry::Entry;
+
+pub use path::{VendoredPath, VendoredPathBuf, VfsPath};
+
+use crate::file_system::{FileRevision, FileSystemPath};
+use crate::{Db, FxDashMap};
+
+mod path;
+
+/// Virtual file system that supports files from different sources.
+///
+/// The [`Vfs`] supports accessing files from:
+///
+/// * The file system
+/// * Vendored files that are part of the distributed Ruff binary
+///
+/// ## Why do both the [`Vfs`] and [`FileSystem`](crate::FileSystem) trait exist?
+///
+/// It would have been an option to define [`FileSystem`](crate::FileSystem) in a way that all its operation accept
+/// a [`VfsPath`]. This would have allowed to unify most of [`Vfs`] and [`FileSystem`](crate::FileSystem). The reason why they are
+/// separate is that not all operations are supported for all [`VfsPath`]s:
+///
+/// * The only relevant operations for [`VendoredPath`]s are testing for existence and reading the content.
+/// * The vendored file system is immutable and doesn't support writing nor does it require watching for changes.
+/// * There's no requirement to walk the vendored typesystem.
+///
+/// The other reason is that most operations know if they are working with vendored or file system paths.
+/// Requiring them to convert the path to an `VfsPath` to test if the file exist is cumbersome.
```

Comment:
I can say from using the API that it's working pretty well so far. But I admit that the terminology is confusing. Do you have any suggestions on how the naming could be improved or what specifically you find confusing.

I can try to improve the documentation. The way I think about FileSystem is less that it is a virtual file system. It's rather more how we access it. Mwhat would you think of FileSystemDriver?

Reply:
I spent some more time looking at it and I think it's fine, if it seems to you to be working out well. I'm not sure that a rename of `FileSystem` to `FileSystemDriver` would make much difference; I think "Driver" ends up being kind of an empty filler word there that doesn't really communicate much.

It was hard for me to figure out the model for how these two things interact, since neither one encapsulates the other. It seems the model is that a Db has both a Vfs and a FileSystem, and it will use the FileSystem for looking up filesystem paths, and manage looking up vendored paths itself, though that hasn't been implemented yet, just stubbed.

Reply:
Maybe what's confusing here is the naming of `Vfs`? Maybe it should just be `VfsFiles`? 

Anyway, I do think your concern is valid and I went back and forth a couple of times between `FileSystem` handling all `VfsPath`s and the `FileSystem` only handling specific paths. And maybe my reasoning that vendored paths are different enough to not pass them through `FileSystem` is unjustified, considering that calling `write_file` with a path pointing to a directory also fails. So there's anyway some extra care that needs to be taken when handling paths that aren't checked at compile time. But I find it kind of nice if we can catch some of them at compile time.

The way I think about it is that `FileSystem` is a replacement for `std::fs`, that's it. It doesn't add support for using multiple file systems at once (vendored and the regular one), which is what `Vfs` does. 

An entirely different design (not fully thought through) would be to ditch the `FileSystem` trait all together. The main motivation for it is that we can support unsaved files and untitled files in the LSP use case. But we could support this in another way by adding a `open_file` and `close_file` function to `Vfs` that allows to manually override the content of a file. 

The only functionality we would looe is that we can't "mock out" the file system during tests with an in memory file system. But maybe that's not worth going through all that trouble. The WASM integration story would then require to provide a "stub" FS implementation. It's not a big detail and something that e.g. emscripten provides out of the box but it comes with a bit more boilerplate code than an option to just use a memory file system. 

Either way, I don't consider the Vfs / FileSystem design that I proposed here as the final design. It requires some more iterating and I'm open to redesigning it completely. 

* We merge what we have now, with the risk that we might need to refactor some of the calling code
* We strip out the `FileSystem` trait and revisit the design when implementing LSP support.

I'm leaning towards keeping what we have here. There's not a lot of downstream code depending on `FileSystem`, so that a refactor should be fairly painless.